src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/tools/cortisol/IngestTask.java

```diff
@@ -0,0 +1,172 @@
+
+                        docs.add(doc);
+                    }
+                    makeBulkRequest(docs);
+                } else {
+                    break;
+                }
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private RestHighLevelClient buildEsClient() {
+        final RestClientBuilder builder = RestClient.builder(new HttpHost(endpoint, port, "http"));
+        return new RestHighLevelClient(builder);
+    }
+
+    private void createIndex(final String mappingTemplateJson, int nPrimaryShards, int nReplicaShards) throws IOException {
+        final CreateIndexRequest cir = new CreateIndexRequest(indexName);
+        cir.mapping(mappingTemplateJson, XContentType.JSON);
+        cir.settings(CortisolHelper.buildIndexSettingsJson(nPrimaryShards, nReplicaShards), XContentType.JSON);
+        final CreateIndexResponse response = client.indices().create(cir, RequestOptions.DEFAULT);
+        assert response.isAcknowledged();
+    }
+
+    private void makeBulkRequest(final List<String> bulkDocs) {
+        final BulkRequest bulkRequest = new BulkRequest();
+        for (String bulkDoc : bulkDocs) {
+            final IndexRequest indexRequest = Requests.indexRequest(indexName);
+            indexRequest.source(bulkDoc, XContentType.JSON);
+            bulkRequest.add(indexRequest);
+        }
+        client.bulkAsync(bulkRequest, RequestOptions.DEFAULT, new ActionListener<BulkResponse>() {
+            @Override
+            public void onResponse(BulkResponse bulkItemResponses) {
+                if (!bulkItemResponses.hasFailures()) {
+                    successCount.addAndGet(bulkDocs.size());
+                } else {
+                    for (BulkItemResponse response : bulkItemResponses) {
+                        if (response.isFailed()) {
+                            failureCount.incrementAndGet();
+                        } else {
+                            successCount.incrementAndGet();
+                        }
+                    }
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                failureCount.addAndGet(bulkDocs.size());
+            }
+        });
+        requestCount.addAndGet(bulkDocs.size());
+    }
+
+    private class TokenBucket {
+        private final Lock lock = new ReentrantLock();
+        private final Condition hasTokens = lock.newCondition();
+        private final ScheduledExecutorService ses = Executors.newSingleThreadScheduledExecutor();
+
+        private final int maxTokens;
+        private int nTokens;
+
+        public TokenBucket(final int maxTokens) {
+            this.maxTokens = maxTokens;
+            this.ses.schedule(this::refill, 1, TimeUnit.SECONDS);
+        }
+
+        public int take() {
+            lock.lock();
+            try {
+                while (nTokens == 0) {
+                    hasTokens.await();
+                }
```

Comment:
Does this need to be in a while loop?

Reply:
Yes, to guard against spurious wake ups.